web/bridge_types_controller.go

```diff
@@ -19,6 +19,10 @@ func (btc *BridgeTypesController) Create(c *gin.Context) {
 		c.JSON(500, gin.H{
 			"errors": []string{err.Error()},
 		})
+	} else if err = services.ValidateAdaptor(bt, btc.App.Store); err != nil {
+		c.JSON(400, gin.H{
```

Comment:
Does this controller method have to worry about concurrency? If so, you should implement locking since you are reading from the bolt database.

Reply:
Reading and writing to bolt db is thread safe so we shouldn’t need locking. Reference material:

https://github.com/boltdb/bolt/blob/master/README.md#project-status

I do think we should use the interface’s method ‘GetStore’ over the field when passing to ValidateAdapter.

Reply:
What I meant is that this controller method is not thread safe. Correct me if I'm wrong, but two threads trying to add a bridge type that doesn't exist , with the same name but different URL can pass through ValidateAdaptor, and then try to write to the database. One of the threads will overwrite the other thread. Both will return 200 as response.

Reply:
Yes I believe you're correct. We'll need to collapse `ValidateAdaptor` and `Save` into a locking method `SaveAdapter` to be thread safe. Want to do the honors or shall I? Either way, thanks a lot @aklintsevich!

Reply:
I'll do it